File: orion-viewer/src/main/java/universe/constellation/orion/viewer/prefs/ActionListActivity.java
/*
 * Orion Viewer - pdf, djvu, xps and cbz file viewer for android devices
 *
 * Copyright (C) 2011-2013  Michael Bogdanov & Co
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package universe.constellation.orion.viewer.prefs;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.CheckedTextView;
import android.widget.ListView;
import android.widget.TextView;

import androidx.annotation.NonNull;

import universe.constellation.orion.viewer.Action;
import universe.constellation.orion.viewer.OrionBaseActivity;
import universe.constellation.orion.viewer.R;


public class ActionListActivity extends OrionBaseActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        onOrionCreate(savedInstanceState, R.layout.actions_selection, false);

        int type = getIntent().getIntExtra("type", 0);
        TextView header = findViewById(R.id.actions_header);
        header.setText(type == 0 ? R.string.short_click : type == 1 ? R.string.long_click : R.string.binding_click);
        final int keyCode = getIntent().getIntExtra("keyCode", 0);
        final boolean isLong = getIntent().getBooleanExtra("isLong", false);
        if (type == 2) {
            String name = KeyEventNamer.getKeyName(keyCode);
            header.setText(header.getText().toString() + " " + name + (isLong ? " [long press]" : ""));
        }

        ListView view = findViewById(R.id.actionsGroup);
        final Action [] actions = Action.values();
        view.setAdapter(new ArrayAdapter<Action>(this, android.R.layout.simple_list_item_single_choice, Action.values()) {
            @NonNull
            @Override
            public View getView(int position, View convertView, @NonNull ViewGroup parent) {
                CheckedTextView view = (CheckedTextView)super.getView(position, convertView, parent);
                view.setText(actions[position].getNameRes());
                return view;
            }
        });

        view.setChoiceMode(ListView.CHOICE_MODE_SINGLE);

        int code = getIntent().getIntExtra("code", 0);

        for (int i = 0; i < actions.length; i++) {
            Action action = actions[i];
            if (action.code == code) {
                view.setItemChecked(i, true);
                //view.setSelection(i);
                break;
            }
        }
        view.setOnItemClickListener((parent, view1, position, id) -> {
            int code1 = actions[position].code;
            Intent result = new Intent();
            result.putExtra("code", code1);
            result.putExtra("keyCode", keyCode);
            result.putExtra("isLong", isLong);
            setResult(Activity.RESULT_OK, result);
            finish();
        });
    }

    public OrionApplication getOrionContext() {
        return (OrionApplication) getApplicationContext();
    }
}

File: orion-viewer/src/main/java/universe/constellation/orion/viewer/prefs/ActionListActivity.kt
package universe.constellation.orion.viewer.prefs

import android.annotation.SuppressLint
import android.graphics.drawable.Drawable
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import android.widget.CheckedTextView
import android.widget.EditText
import android.widget.ListView
import android.widget.TextView
import androidx.annotation.DrawableRes
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.graphics.drawable.DrawableCompat
import androidx.core.widget.doAfterTextChanged
import universe.constellation.orion.viewer.Action
import universe.constellation.orion.viewer.ActionGroup
import universe.constellation.orion.viewer.OrionBaseActivity
import universe.constellation.orion.viewer.R

/**
 * Picks the action of a tap zone or a key: "None" first, then every action under the header of
 * its group. A header tap folds the group; a query in the search box shows the matching actions
 * of every group instead.
 */
class ActionListActivity : OrionBaseActivity() {

    private val collapsed = HashSet<ActionGroup>()

    private var query = ""

    private var selected: Action? = null

    private val rows = ArrayList<Any>()

    @SuppressLint("MissingSuperCall")
    override fun onCreate(savedInstanceState: Bundle?) {
        onOrionCreate(savedInstanceState, R.layout.actions_selection, false)

        val type = intent.getIntExtra("type", 0)
        val header = findViewById<TextView>(R.id.actions_header)
        header.setText(if (type == 0) R.string.short_click else if (type == 1) R.string.long_click else R.string.binding_click)
        val keyCode = intent.getIntExtra("keyCode", 0)
        val isLong = intent.getBooleanExtra("isLong", false)
        if (type == 2) {
            val name = KeyEventNamer.getKeyName(keyCode)
            header.text =
                header.text.toString() + " " + name + (if (isLong) " [long press]" else "")
        }

        val code = intent.getIntExtra("code", 0)
        selected = Action.entries.firstOrNull { it.code == code }
        //ten rows for hardware keys of e-ink readers: folded unless one of them is the choice
        if (selected?.group != ActionGroup.CROP_KEYS) collapsed.add(ActionGroup.CROP_KEYS)

        val view = findViewById<ListView>(R.id.actionsGroup)
        val adapter = object : ArrayAdapter<Any>(this, android.R.layout.simple_list_item_single_choice, rows) {
            override fun getViewTypeCount() = 2

            override fun getItemViewType(position: Int) = if (rows[position] is ActionGroup) 1 else 0

            override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
                return when (val row = rows[position]) {
                    is ActionGroup -> {
                        val header = (convertView as? TextView)
                            ?: (layoutInflater.inflate(android.R.layout.preference_category, parent, false) as TextView).apply {
                                compoundDrawablePadding = resources.getDimensionPixelSize(R.dimen.action_group_icon_padding)
                            }
                        header.setText(row.titleRes)
                        //the group icon at the start, a chevron at the end while the group is folded
                        val folded = row in collapsed && query.isEmpty()
                        header.setCompoundDrawablesRelative(
                            headerIcon(row.iconRes, header),
                            null,
                            if (folded) headerIcon(R.drawable.outline_chevron_right, header) else null,
                            null
                        )
                        header
                    }
                    else -> {
                        val item = (convertView as? CheckedTextView)
                            ?: layoutInflater.inflate(android.R.layout.simple_list_item_single_choice, parent, false) as CheckedTextView
                        item.setText((row as Action).nameRes)
                        item.isChecked = row == selected
                        item
                    }
                }
            }
        }
        view.adapter = adapter
        rebuildRows()

        view.onItemClickListener = OnItemClickListener { _: AdapterView<*>?, _: View?, position: Int, _: Long ->
            when (val row = rows[position]) {
                is ActionGroup -> {
                    if (query.isEmpty()) {
                        if (!collapsed.remove(row)) collapsed.add(row)
                        rebuildRows()
                    }
                }
                is Action -> {
                    val result = Intent()
                    result.putExtra("code", row.code)
                    result.putExtra("keyCode", keyCode)
                    result.putExtra("isLong", isLong)
                    setResult(RESULT_OK, result)
                    finish()
                }
            }
        }

        findViewById<EditText>(R.id.actions_search).doAfterTextChanged {
            query = it?.toString()?.trim() ?: ""
            rebuildRows()
        }
    }

    /** A header icon at the size and in the colour of the header text. */
    private fun headerIcon(@DrawableRes res: Int, header: TextView): Drawable? {
        val drawable = AppCompatResources.getDrawable(this, res) ?: return null
        val size = resources.getDimensionPixelSize(R.dimen.action_group_icon_size)
        return DrawableCompat.wrap(drawable.mutate()).apply {
            setBounds(0, 0, size, size)
            DrawableCompat.setTintList(this, header.textColors)
        }
    }

    /** The list for the current query and folding; the adapter shares [rows]. */
    private fun rebuildRows() {
        rows.clear()
        if (matches(Action.NONE)) rows.add(Action.NONE)
        for (group in ActionGroup.entries) {
            val actions = Action.entries.filter { it.group == group && matches(it) }
            if (actions.isEmpty()) continue
            rows.add(group)
            if (query.isNotEmpty() || group !in collapsed) rows.addAll(actions)
        }
        (findViewById<ListView>(R.id.actionsGroup).adapter as ArrayAdapter<*>).notifyDataSetChanged()
    }

    private fun matches(action: Action): Boolean =
        query.isEmpty() || getString(action.nameRes).contains(query, ignoreCase = true)
}
